### PR TITLE
Amp 56383 go sdk docs

### PR DESCRIPTION
**Pull request by Mercy811**
# Dev Docs PR

## Description
Adds docs for Go SDK (Alpha) for Analytics.

### Update Go SDK (Alpha) docs
Screen Shot 2022-08-02 at 21 58 48

### Updated SDK sources list
Screen Shot 2022-08-02 at 21 59 35

## Deadline

When do these changes need to be live on the site?

## Change type

- [ ] Doc bug fix. Fixes #[insert issue number]. Amplitude contributors include Jira issue number. 
- [ ] Doc update.
- [X] New documentation.
- [ ] Non-documentation related fix or update.

# PR checklist:

- [x] My documentation follows the style guidelines of this project.
- [x] I previewed my documentation on a local server using `mkdocs serve`.
- [x] Running `mkdocs serve` didn't generate any failures.
- [x] I have performed a self-review of my own documentation.

### CODE REVIEW

**Review by caseyamp — APPROVED** [review]

**Review by caseyamp — APPROVED** [review]

File: docs/data/sdks/go.md [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+title: Go SDK (Alpha)
+description: The Amplitude Go SDK installation and quick start guide.
+icon: fontawesome/brands/golang
+---
+
+!!!info "SDK Resources"
+    - [Go SDK Repository :material-github:](https://github.com/amplitude/Amplitude-Go)
+    - [Go SDK Releases :material-code-tags-check:](https://github.com/amplitude/Amplitude-Go/releases)
+
+--8<-- "includes/ampli-vs-amplitude.md"
```
Comment (kevinpagtakhan) [inline docs/data/sdks/go.md:11]:
```suggestion
--8<-- "includes/no-ampli.md"
```

File: docs/data/sdks/go.md [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+	//code: status code of request response
+	//message: a optional string message for more detailed information
+}
+
+Config{
+		APIKey:          "your-api-key",
+		FlushInterval:   time.Second * 10,
+		FlushQueueSize:  100,
+		FlushMaxRetries: 5,
+		Logger:          newDefaultLogger(),
```
Comment (kevinpagtakhan) [inline docs/data/sdks/go.md:63]: Do customer have access to `newDefaultLogger()`?

File: docs/data/sdks/go.md [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+		MinIDLength:     7,
+		Callback:        callback,
+		ServerZone:      "EU",
+		UseBatch:        true,
+		OptOut:          false,
+		ServerURL:       "proxy url that forwarding the requests",
```
Comment (kevinpagtakhan) [inline docs/data/sdks/go.md:69]: Let's use valid URL formatted string here; otherwise this is confusing

File: docs/data/sdks/go.md [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+
+Config{
+		APIKey:          "your-api-key",
+		FlushInterval:   time.Second * 10,
+		FlushQueueSize:  100,
+		FlushMaxRetries: 5,
+		Logger:          newDefaultLogger(),
+		MinIDLength:     7,
+		Callback:        callback,
+		ServerZone:      "EU",
```
Comment (kevinpagtakhan) [inline docs/data/sdks/go.md:66]: Is this not in an enum?

File: docs/data/sdks/go.md [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+Set your configuration before a client is initialized.
+```Go
+func callback(event Event, code int, message ...string) {
+	//callback function that takes three input parameters
+	//event: the event that triggered this callback
+	//code: status code of request response
+	//message: a optional string message for more detailed information
+}
+
+Config{
```
Comment (kevinpagtakhan) [inline docs/data/sdks/go.md:58]: Are all of these properties required to be defined or are some optional?

File: docs/data/sdks/go.md [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+client.Track(event)
+
+// Track events with optional properties
+client.Track(amplitude.Event{
+		EventType:       "type-of-event",
+		EventOptions:    amplitude.EventOptions{
+			UserID:             "user-id",
+			DeviceID:           "device-id",
+		},
+		EventProperties: map[string]interface{}{"source": "notification"},
```
Comment (kevinpagtakhan) [inline docs/data/sdks/go.md:109]: Any chance we dont need to add `map[string]interface{}`? or have something shorter?

File: docs/data/sdks/go.md [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+
+The `Flush` method triggers the client to send buffered events.
+
+```Go
+client.Flush()
+```
+
+### `Add`
+
+The `Add` method adds a plugin to Amplitude client struct. Plugins can help processing and sending events. [Learn more about plugins.](#amplitude-sdk-plugin).
```
Comment (kevinpagtakhan) [inline docs/data/sdks/go.md:300]:
```suggestion
The `Add` method adds a plugin to Amplitude client struct. Plugins can help processing and sending events. [Learn more about plugins](#amplitude-sdk-plugin).
```

File: docs/data/sdks/go.md [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+This method contains the logic for processing events and has `*Event` as parameter. If used as enrichment type plugin, the expected return value is the modified/enriched event; while if used as a destination type plugin, the expected return value is `nil`. This method is called for each event, including Identify, GroupIdentify and Revenue events, that is instrumented using the client interface.
+
+### Plugin Examples
+
+#### Enrichment Type Plugin
+
+Here's an example of a plugin that modifies each event that is instrumented by adding an increment integer to `EventId` property of an event.
+
+```Go
+type addEventPlugin struct {
```
Comment (kevinpagtakhan) [inline docs/data/sdks/go.md:341]:
```suggestion
type addEventIdPlugin struct {
```

File: docs/data/sdks/go.md [hunk 9]
```diff
@@ -0,0 +1,405 @@
...
+	currentID int
+	config    amplitude.Config
+}
+
+func (plugin *addEventPlugin) Setup(config amplitude.Config) {
+	plugin.config = config
+}
+
+func (plugin *addEventPlugin) Priority() amplitude.EnrichmentPriority {
```
Comment (kevinpagtakhan) [inline docs/data/sdks/go.md:350]: Priority?

File: docs/data/sdks/go.md [hunk 10]
```diff
@@ -0,0 +1,405 @@
...
+	url string
+	config amplitude.Config
+}
+
+func (plugin *myDestinationPlugin) Setup(config amplitude.Config) {
+	plugin.config = config
+}
+
+func (plugin *myDestinationPlugin) Execute(event *amplitude.Event) {
+	payload := map[string]interface{}{"key": "secret", "event":event}
```
Comment (kevinpagtakhan) [inline docs/data/sdks/go.md:378]:
```suggestion
	payload := map[string]interface{}{"key": "secret", "events":event}
```

Comment (kevinpagtakhan) [reply docs/data/sdks/go.md:378]: Also events payload should be an array

**Comment by kevinpagtakhan** [comment]
@bgiori when you get the chance, can you take a look at the code sample also? curious if the usage is the ideal DX

**Review by kevinpagtakhan — CHANGES_REQUESTED** [review]
See comments above

File: docs/data/sdks/go.md (hunk 3 above)
Comment (Mercy811) [reply docs/data/sdks/go.md:69]: updated it to "https://api2.amplitude.com/2/httpapi" the default server url

File: docs/data/sdks/go.md (hunk 2 above)
Comment (Mercy811) [reply docs/data/sdks/go.md:63]: deleted this line

File: docs/data/sdks/go.md (hunk 4 above)
Comment (Mercy811) [reply docs/data/sdks/go.md:66]: Changed it to enum `ServerZoneEU`

File: docs/data/sdks/go.md (hunk 2 above)
Comment (Mercy811) [reply docs/data/sdks/go.md:63]: Not sure which fields you need here

File: docs/data/sdks/go.md (hunk 5 above)
Comment (Mercy811) [reply docs/data/sdks/go.md:58]: Users should use `NewConfig(apiKey)` to init a config struct. Not sure if I should delete this block of code here

File: docs/data/sdks/go.md (hunk 6 above)
Comment (Mercy811) [reply docs/data/sdks/go.md:109]: It's required to match data types. `EventProperties`'s data type is `map[string]interface{}`. I guess no way to get rid of it

File: docs/data/sdks/go.md (hunk 10 above)
Comment (Mercy811) [reply docs/data/sdks/go.md]: It should be just a single event. Typescript example https://www.docs.developers.amplitude.com/data/sdks/typescript-node/#destination-type-plugin

File: docs/data/sdks/go.md (hunk 9 above)
Comment (Mercy811) [reply docs/data/sdks/go.md:350]: Changed it to `Type()`. This doc finished earlier than Go Alpha, forgot to change doc.

File: docs/data/sdks/go.md [hunk 11]
```diff
@@ -0,0 +1,404 @@
...
+
+## Usage
+
+### Initializing SDK
+
+Initialization is necessary before any instrumentation is done. The API key for your Amplitude project is required to construct a Config struct. Use that Config struct to initialize a client struct which implements Client interface. It can be used across requests after it is initialized.
+
+```GO
+
+import "github.com/amplitude/Amplitude-Go/amplitude"
```
Comment (kevinpagtakhan) [inline docs/data/sdks/go.md:33]: replace name to the new name

File: docs/data/sdks/go.md [hunk 12]
```diff
@@ -0,0 +1,404 @@
...
+
+func main() {
+
+    // Create a Config struct 
+	config := amplitude.NewConfig("your-api-key")
+
+	// Pass a Config struct  
+    // to initialize a client struct
+    // which implements Client interface
+	client := amplitude.NewClient(config)
```
Comment (kevinpagtakhan) [inline docs/data/sdks/go.md:43]: let's fix the indentation

File: docs/data/sdks/go.md [hunk 13]
```diff
@@ -0,0 +1,404 @@
...
+}
+
+Config{
+		APIKey:          "your-api-key",
+		FlushInterval:   time.Second * 10,
+		FlushQueueSize:  100,
+		FlushMaxRetries: 5,
+		MinIDLength:     7,
+		Callback:        callback,
+		ServerZone:      ServerZoneEU,
```
Comment (kevinpagtakhan) [inline docs/data/sdks/go.md:65]:
```suggestion
		ServerZone:      amplitude.ServerZoneEU,
```

File: docs/data/sdks/go.md [hunk 14]
```diff
@@ -0,0 +1,404 @@
...
+		UseBatch:        true,
+		OptOut:          false,
+		ServerURL:       HTTPV2,
```
Comment (kevinpagtakhan) [inline docs/data/sdks/go.md:68]:
```suggestion
		ServerURL:       amplitude.HTTPV2,
```

File: docs/data/sdks/go.md (hunk 13 above)
Comment (kevinpagtakhan) [reply docs/data/sdks/go.md:65]: let's remove this in the docs for now since it's not supported

File: docs/data/sdks/go.md [hunk 15]
```diff
@@ -0,0 +1,404 @@
...
+
+Config{
+		APIKey:          "your-api-key",
+		FlushInterval:   time.Second * 10,
+		FlushQueueSize:  100,
+		FlushMaxRetries: 5,
+		MinIDLength:     7,
+		Callback:        callback,
+		ServerZone:      ServerZoneEU,
+		UseBatch:        true,
```
Comment (kevinpagtakhan) [inline docs/data/sdks/go.md:66]: let's also remove this

File: docs/data/sdks/go.md [hunk 16]
```diff
@@ -0,0 +1,404 @@
...
+	//message: a optional string message for more detailed information
+}
+
+Config{
+		APIKey:          "your-api-key",
+		FlushInterval:   time.Second * 10,
+		FlushQueueSize:  100,
+		FlushMaxRetries: 5,
+		MinIDLength:     7,
+		Callback:        callback,
```
Comment (kevinpagtakhan) [inline docs/data/sdks/go.md:64]: let's also remove this

File: docs/data/sdks/go.md (hunk 14 above)
Comment (kevinpagtakhan) [inline docs/data/sdks/go.md:68]: Let's pass string url

File: docs/data/sdks/go.md [hunk 17]
```diff
@@ -0,0 +1,404 @@
...
+		},
+		EventProperties: map[string]interface{}{"source": "notification"},
+})
+```
+
+### User Properties
+
+User properties help you understand your users at the time they perform some action within your app such as their device details, their preferences, or language.
+
+Identify is for setting the user properties of a particular user without sending any event. The SDK supports the operations `Set`, `SetOnce`, `Unset`, `Add`, `Append`, `Prepend`, `PreInsert`, `PostInsert`,`Remove`, and `ClearAll1` on individual user properties. The operations are declared as Identify struct methods. Multiple operations can be chained together in a single Identify struct. The Identify struct is then passed to the Amplitude client to send to the server.
```
Comment (kevinpagtakhan) [inline docs/data/sdks/go.md:116]:
```suggestion
Identify is for setting the user properties of a particular user without sending any event. The SDK supports the operations `Set`, `SetOnce`, `Unset`, `Add`, `Append`, `Prepend`, `PreInsert`, `PostInsert`,`Remove`, and `ClearAll` on individual user properties. The operations are declared as Identify struct methods. Multiple operations can be chained together in a single Identify struct. The Identify struct is then passed to the Amplitude client to send to the server.
```

File: docs/data/sdks/go.md [hunk 18]
```diff
@@ -0,0 +1,404 @@
...
+	//event: the event that triggered this callback
+	//code: status code of request response
+	//message: a optional string message for more detailed information
+}
+
+Config{
+		APIKey:          "your-api-key",
+		FlushInterval:   time.Second * 10,
+		FlushQueueSize:  100,
+		FlushMaxRetries: 5,
```
Comment (kevinpagtakhan) [inline docs/data/sdks/go.md:62]: let's remove this in code as well

File: docs/data/sdks/go.md (hunk 11 above)
Comment (Mercy811) [reply docs/data/sdks/go.md]: replace it with github.com/amplitude/analytics-go

File: docs/data/sdks/go.md (hunk 13 above)
Comment (Mercy811) [reply docs/data/sdks/go.md]: fixed

File: docs/data/sdks/go.md (hunk 12 above)
Comment (Mercy811) [reply docs/data/sdks/go.md]: fixed

File: docs/data/sdks/go.md (hunk 14 above)
Comment (Mercy811) [reply docs/data/sdks/go.md]: Use `NewConfig()` in current example.

**Review by bohan-amplitude — APPROVED** [review]
LGTM

File: docs/data/sdks/go.md (hunk 2 above)
Comment (Mercy811) [reply docs/data/sdks/go.md]: Deleted all fields unused in Alpha version

File: docs/data/sdks/go.md (hunk 5 above)
Comment (Mercy811) [reply docs/data/sdks/go.md]: Update example [here](https://pr-189.d19s7xzcva2mw3.amplifyapp.com/data/sdks/go/#configuration). Use `NewConfig` to init a config

File: docs/data/sdks/go.md (hunk 15 above)
Comment (Mercy811) [reply docs/data/sdks/go.md]: Deleted all fields unused in Alpha version

File: docs/data/sdks/go.md (hunk 16 above)
Comment (Mercy811) [reply docs/data/sdks/go.md]: Deleted all fields unused in Alpha version

File: docs/data/sdks/go.md (hunk 18 above)
Comment (Mercy811) [reply docs/data/sdks/go.md]: Deleted all fields unused in Alpha version

File: docs/data/sdks/go.md [hunk 19]
```diff
@@ -0,0 +1,403 @@
...
+
+Set your configuration before a client is initialized.
+
+| <div class="big-column">Name</div> | Description  |
+| --- | --- |
+| `APIKey` | Required. string. The API key of the Amplitude project. Events sent by the client struct can be found under this project. Set when you initialize the client struct. |
+| `FlushQueueSize` | int. Events wait in the buffer and are sent in a batch. The buffer is flushed when the number of events reaches `FlushQueueSize`. Defaults to 200.|
+| `FlushInterval` | time.Duration. Events wait in the buffer and are sent in a batch. The buffer is flushed every `FlushInterval`. Defaults to 10 seconds.|
+| `Logger` | Logger interface. The logger used by Amplitude client. Defaults to using a wrapper of [Go standard Logger](https://pkg.go.dev/log#Logger): `log.Logger`. |
+| `ServerURL` | string. The API endpoint URL that events are sent to. Automatically selected by `ServerZone` and `UseBatch`. If this field is set with a string value instead of `nil`, then `ServerZone` and `UseBatch` are ignored and the string value is used. Defaults to the HTTP API V2 endpoint. |
```
Comment (kevinpagtakhan) [inline docs/data/sdks/go.md:62]:
```suggestion
| `ServerURL` | string. The API endpoint URL that events are sent to. |
```

File: docs/data/sdks/go.md [hunk 20]
```diff
@@ -0,0 +1,403 @@
...
+| `Storage` | Storage interface. Used to create storage struct to hold events in the storage buffer. Events in storage buffer are waiting to be sent. Defaults to `InMemoryStorage`. |
+| `OptOut`  | bool. Opt out option. If set to `true`, client doesn't process and send events. Defaults to `false`. |
+
+### Track an event
+
+Events represent how users interact with your application. For example, “Button Clicked” may be an action you want to note.
```
Comment (kevinpagtakhan) [inline docs/data/sdks/go.md:68]:
```suggestion
Events represent how users interact with your application. For example, "Button Clicked" may be an action you want to note.
```

File: docs/data/sdks/go.md [hunk 21]
```diff
@@ -0,0 +1,403 @@
...
+	if err != nil{
+		plugin.config.Logger.Error("Event encoding failed: ", err)
+	}
+
+	request, err := http.NewRequest("POST", plugin.config.ServerURL, bytes.NewReader(payloadBytes))
+	if err != nil {
+		plugin.config.Logger.Error("Building new request failed", err)
+	}
+
+	httpClient := http.Client{}
```
Comment (kevinpagtakhan) [inline docs/data/sdks/go.md:385]: Lets move this to setup

**Review by kevinpagtakhan — APPROVED** [review]

**Comment by Mercy811** [comment]
The above changes are meant to resolve conflicts with main branch. It should be the correct version and thanks @liuyang1520 for helping.

Try to explain what was going on:
- somehow I forked this repo and git origin was set to the forked repo
- I forgot origin was set to forked repo and rebased branch to main

**Pull request merged**